[Perf] Async Scheduling + Speculative Decoding + Structured Outputs

benchislett opened the pull request:
## Purpose

This PR enables structured outputs when using speculative decoding and async scheduling. The solution implemented in this PR is as follows:
- After computing the draft tokens, asynchronously copy them to the CPU. Overlap this copy with the next model forward step, on a separate stream.
- Concurrently, after calling the next `execute_model`, the scheduler will call `take_draft_token_ids` which will wait for the draft tokens to arrive on the cpu and yield them. 
- The scheduler uses the draft tokens to generate the bitmask for the batch (including spec tokens. still concurrent with execute_model), and then calls sample_tokens as before.

In order to accommodate for invalid draft tokens (which do not adhere to the schema), we replace them on the scheduler_output object and pad the rejected tokens with `-1`, which will be ignored when filling the bitmasks. To ensure that the model does not sample from a position after an invalid token, we attach a per-request count of invalid tokens onto the grammar output to mask out the sampled positions.

## Test Plan

GSM8K to check for base-model correctness regressions.

Structured outputs `xgrammar_bench` for speed, and also with `json_unique` for grammar adherence.

## Test Result

Passes correctness and structured output adherence tests.

Benchmarking details:

### Setup:
```
vllm serve meta-llama/Llama-3.1-8B-Instruct --speculative-config '{"method": "eagle3", "model": "yuhuili/EAGLE3-LLaMA3.1-Instruct-8B", "num_speculative_tokens": 3}' --async-scheduling --max-num-seqs 128 --no-enable-prefix-caching
```
Three runs: one with async + spec, one with only spec, one with only async.

 ### Result:
100% coverage of xgrammar_bench in all cases.
Perf measurement (when controlling for output tokens with ignore-eos):

#### With Structured Outputs

- Async Scheduling **(1.00x E2E)**
```
============ Serving Benchmark Result ============
Successful requests:                     100       
Maximum request concurrency:             10        
Benchmark duration (s):                  4.71      
Total input tokens:                      30919     
Total generated tokens:                  9569      
Request throughput (req/s):              21.25     
Output token throughput (tok/s):         2033.52   
Total Token throughput (tok/s):          8604.16   
---------------Time to First Token----------------
Mean TTFT (ms):                          46.46     
Median TTFT (ms):                        49.31     
P99 TTFT (ms):                           70.04     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          4.44      
Median TPOT (ms):                        4.42      
P99 TPOT (ms):                           4.73      
---------------Inter-token Latency----------------
Mean ITL (ms):                           4.25      
Median ITL (ms):                         4.08      
P99 ITL (ms):                            10.61     
==================================================
```

- Spec Decoding **(1.38x)**
```
============ Serving Benchmark Result ============
Successful requests:                     100       
Maximum request concurrency:             10        
Benchmark duration (s):                  3.40      
Total input tokens:                      30919     
Total generated tokens:                  9574      
Request throughput (req/s):              29.41     
Output token throughput (tok/s):         2815.90   
Total Token throughput (tok/s):          11909.78  
---------------Time to First Token----------------
Mean TTFT (ms):                          27.96     
Median TTFT (ms):                        23.60     
P99 TTFT (ms):                           71.90     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          3.11      
Median TPOT (ms):                        3.14      
P99 TPOT (ms):                           4.10      
---------------Inter-token Latency----------------
Mean ITL (ms):                           6.97      
Median ITL (ms):                         6.35      
P99 ITL (ms):                            13.68     
==================================================
```

- Async Scheduling + Spec Decoding **(1.62x)**
```
============ Serving Benchmark Result ============
Successful requests:                     100       
Maximum request concurrency:             10        
Benchmark duration (s):                  2.91      
Total input tokens:                      30919     
Total generated tokens:                  9570      
Request throughput (req/s):              34.32     
Output token throughput (tok/s):         3284.67   
Total Token throughput (tok/s):          13896.85  
---------------Time to First Token----------------
Mean TTFT (ms):                          35.58     
Median TTFT (ms):                        31.50     
P99 TTFT (ms):                           71.67     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          2.56      
Median TPOT (ms):                        2.60      
P99 TPOT (ms):                           3.52      
---------------Inter-token Latency----------------
Mean ITL (ms):                           5.76      
Median ITL (ms):                         5.09      
P99 ITL (ms):                            14.75     
==================================================
```

#### No Structured Outputs

- Async Scheduling **(1.00x E2E)**
```
============ Serving Benchmark Result ============
Successful requests:                     100       
Maximum request concurrency:             10        
Benchmark duration (s):                  4.55      
Total input tokens:                      30919     
Total generated tokens:                  9645      
Request throughput (req/s):              21.99     
Output token throughput (tok/s):         2121.08   
Total Token throughput (tok/s):          8920.62   
---------------Time to First Token----------------
Mean TTFT (ms):                          39.46     
Median TTFT (ms):                        41.62     
P99 TTFT (ms):                           56.57     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          4.34      
Median TPOT (ms):                        4.33      
P99 TPOT (ms):                           4.69      
---------------Inter-token Latency----------------
Mean ITL (ms):                           4.18      
Median ITL (ms):                         4.08      
P99 ITL (ms):                            5.13      
==================================================
```

- Spec Decoding **(1.56x)**
```
============ Serving Benchmark Result ============
Successful requests:                     100       
Maximum request concurrency:             10        
Benchmark duration (s):                  2.91      
Total input tokens:                      30919     
Total generated tokens:                  9661      
Request throughput (req/s):              34.42     
Output token throughput (tok/s):         3324.92   
Total Token throughput (tok/s):          13965.99  
---------------Time to First Token----------------
Mean TTFT (ms):                          24.56     
Median TTFT (ms):                        21.39     
P99 TTFT (ms):                           51.44     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          2.66      
Median TPOT (ms):                        2.69      
P99 TPOT (ms):                           3.68      
---------------Inter-token Latency----------------
Mean ITL (ms):                           6.21      
Median ITL (ms):                         5.52      
P99 ITL (ms):                            13.48     
==================================================
```

- Async Scheduling + Spec Decoding **(1.66x)**
```
============ Serving Benchmark Result ============
Successful requests:                     100       
Maximum request concurrency:             10        
Benchmark duration (s):                  2.74      
Total input tokens:                      30919     
Total generated tokens:                  9661      
Request throughput (req/s):              36.45     
Output token throughput (tok/s):         3521.31   
Total Token throughput (tok/s):          14790.89  
---------------Time to First Token----------------
Mean TTFT (ms):                          30.94     
Median TTFT (ms):                        29.38     
P99 TTFT (ms):                           53.58     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          2.43      
Median TPOT (ms):                        2.44      
P99 TPOT (ms):                           3.31      
---------------Inter-token Latency----------------
Mean ITL (ms):                           5.68      
Median ITL (ms):                         5.03      
P99 ITL (ms):                            15.35     
==================================================
```

benchislett committed:
WIP

Signed-off-by: Benjamin Chislett <bchislett@nvidia.com>

benchislett committed:
WIP - functional

Signed-off-by: Benjamin Chislett <bchislett@nvidia.com>

benchislett committed:
revert changes to processor.py

Signed-off-by: Benjamin Chislett <bchislett@nvidia.com>

benchislett committed:
Merge branch 'main' into async-spec-struct

benchislett committed:
update input_processor.py

Signed-off-by: Benjamin Chislett <bchislett@nvidia.com>

benchislett committed:
refactor some more

Signed-off-by: Benjamin Chislett <bchislett@nvidia.com>

benchislett committed:
tidy

Signed-off-by: Benjamin Chislett <bchislett@nvidia.com>

benchislett committed:
Merge branch 'main' into async-spec-struct

njhill commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -1231,6 +1250,7 @@ def _prepare_input_ids(
         )
 
         # Scatter the draft tokens after the sampled tokens are scattered.
+        self._prev_draft_token_ids = self._draft_token_ids
```
I think this is redundant/leftover?

njhill commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1291,14 +1294,25 @@ def update_draft_token_ids(
+                assert metadata is not None and metadata.grammar is not None
+                spec_token_ids = metadata.grammar.validate_tokens(
+                    spec_token_ids,
                 )
```
nit: should fit now?
```suggestion
                spec_token_ids = metadata.grammar.validate_tokens(spec_token_ids)
```

njhill commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -3132,6 +3154,29 @@ def sample_tokens(
+            col_indices = torch.arange(num_sampled_toks, dtype=torch.int32)
+            mask_start_indices = num_sampled_toks - num_invalid_spec_tokens_cpu
+            mask = col_indices.unsqueeze(0) >= mask_start_indices.unsqueeze(1)
+            self.invalid_spec_tokens_mask.cpu[:num_reqs, :].copy_(mask)
```
Should be faster to do this in numpy
```suggestion
            num_invalid = np.zeros((num_reqs,), dtype=np.int32)
            for req_id, num_invalid_toks in zip(
                grammar_output.structured_output_request_ids, num_invalid_tokens_per_req
            ):
                req_index = self.input_batch.req_id_to_index[req_id]
                num_invalid[req_index] = num_invalid_toks
            mask_start_indices = np.subtract(num_sampled_toks, num_invalid, out=num_invalid)
            np.greater_equal(
                np.arange(num_sampled_toks, dtype=np.int32)[None, :],
                mask_start_indices[:, None],
                out=self.invalid_spec_tokens_mask.np[:num_reqs, :]
            )
```

njhill reviewed (commented):
Thanks @benchislett! I am still fully wrapping my head around some of it but I think it looks good mostly.

While reviewing/testing this I found a couple of related issues that we should probably investigate separately (at least the first one):
- There appears to be an existing issue with the combo of structured outputs + preemption + spec decoding `input_fits_in_drafter=False` case (irrespective of async scheduling)
- IIUC, in non-async structured outputs + spec decoding case, there can be a variable number of draft tokens based on how many are accepted by the grammar. I found that this significantly impacts performance because uniform decode is much faster, so wondering if we should pad the grammar-rejected tokens instead

njhill commented:
@benchislett I was running more tests with this, basically extending the cases covered in https://github.com/vllm-project/vllm/blob/main/tests/v1/e2e/test_async_scheduling.py, and found that it doesn't work with:

- Prefill chunking, results in the following crash:
```
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [backend_xgrammar.py:158] Failed to advance FSM for request 96 for tokens 1627. Please file an issue.
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [backend_xgrammar.py:158] Failed to advance FSM for request 95 for tokens 914. Please file an issue.
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [backend_xgrammar.py:158] Failed to advance FSM for request 92 for tokens 1313. Please file an issue.
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [backend_xgrammar.py:158] Failed to advance FSM for request 90 for tokens 508. Please file an issue.
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873] EngineCore encountered a fatal error.
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873] Traceback (most recent call last):
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]   File "/home/nickhill/workspace/vllm2/vllm/vllm/v1/engine/core.py", line 864, in run_engine_core
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]     engine_core.run_busy_loop()
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]   File "/home/nickhill/workspace/vllm2/vllm/vllm/v1/engine/core.py", line 891, in run_busy_loop
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]     self._process_engine_step()
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]   File "/home/nickhill/workspace/vllm2/vllm/vllm/v1/engine/core.py", line 920, in _process_engine_step
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]     outputs, model_executed = self.step_fn()
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]                               ^^^^^^^^^^^^^^
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]   File "/home/nickhill/workspace/vllm2/vllm/vllm/v1/engine/core.py", line 443, in step_with_batch_queue
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]     engine_core_outputs = self.scheduler.update_from_output(
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]   File "/home/nickhill/workspace/vllm2/vllm/vllm/v1/core/sched/scheduler.py", line 1116, in update_from_output
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]     new_token_ids, stopped = self._update_request_with_output(
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]   File "/home/nickhill/workspace/vllm2/vllm/vllm/v1/core/sched/async_scheduler.py", line 61, in _update_request_with_output
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]     assert request.num_output_placeholders >= 0
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(EngineCore_DP0 pid=1560978) ERROR 12-18 03:04:39 [core.py:873] AssertionError
```

- The `input_fits_in_drafter=False` case causes a large number of these log messages:
```
(EngineCore_DP0 pid=1560071) ERROR 12-18 03:04:11 [backend_xgrammar.py:158] Failed to advance FSM for request 70 for tokens 95025. Please file an issue.
(EngineCore_DP0 pid=1560071) ERROR 12-18 03:04:11 [backend_xgrammar.py:158] Failed to advance FSM for request 83 for tokens 334. Please file an issue.
(EngineCore_DP0 pid=1560071) ERROR 12-18 03:04:11 [backend_xgrammar.py:158] Failed to advance FSM for request 70 for tokens 794. Please file an issue.
(EngineCore_DP0 pid=1560071) ERROR 12-18 03:04:11 [backend_xgrammar.py:158] Failed to advance FSM for request 72 for tokens 2627. Please file an issue.
(EngineCore_DP0 pid=1560071) ERROR 12-18 03:04:11 [backend_xgrammar.py:158] Failed to advance FSM for request 79 for tokens 970. Please file an issue.
```
and incorrect output (though it doesn't crash).

benchislett committed:
Merge branch 'main' into async-spec-struct

Signed-off-by: Benjamin Chislett <bchislett@nvidia.com>

benchislett committed:
small patch

Signed-off-by: Benjamin Chislett <bchislett@nvidia.com>

njhill committed:
rework/simplify, fix remaining issues

- revert some unrelated changes
- separate update_draft_tokens_in_output() method
- avoid propagating invalid_spec_tokens separately
- change fits_in_drafter=False case to propose zeros
- fix chunked prefill case by trimming draft tokens
- fix acceptance rate metric to take grammar-rejected tokens into account
- use separate-stream draft token cpu copy for non-async sched case too
- only copy draft tokens to cpu when needed

Signed-off-by: njhill <nickhill123@gmail.com>

njhill commented:
@benchislett I iterated a bit on this in a separate branch: https://github.com/njhill/vllm/tree/async-spec-struct

I've rebased, incorporated all of the above comments, simplified a bit and fixed the prefill chunking issue. Please lmk what you think.

Still remaining:
- Avoid copying draft tokens to cpu in async + no struct output reqs case
- Fix the `input_fits_in_drafter=False` issue

I think those won't take long and then it should be ready from my pov.

njhill committed:
group draft_token_ids_cpu setting logic in same method

Signed-off-by: njhill <nickhill123@gmail.com>

njhill committed:
Merge remote-tracking branch 'refs/remotes/origin/main' into async-spec-struct

Signed-off-by: njhill <nickhill123@gmail.com>

# Conflicts:
#	vllm/v1/core/sched/interface.py
#	vllm/v1/core/sched/scheduler.py
#	vllm/v1/structured_output/__init__.py

benchislett committed:
Merge branch 'main' into async-spec-struct

njhill committed:
precommit and docs fix

Signed-off-by: njhill <nickhill123@gmail.com>

benchislett committed:
Merge remote-tracking branch 'njhill/async-spec-struct' into async-spec-struct

njhill reviewed (approved):
Thanks @benchislett for the teamwork!

benchislett committed:
Merge remote-tracking branch 'upstream/main' into async-spec-struct
